nucypher/policy/policies.py

```diff
@@ -616,19 +751,14 @@ def sample_essential(self,
                 new_to_check = reservoir.draw_at_most(quantity_remaining)
                 to_check.update(new_to_check)
 
-            # Feed newly sampled stakers to the learner
-            self.alice.learn_about_specific_nodes(new_to_check)
-
-            # TODO: would be nice to wait for the learner to finish an iteration here,
-            # because if it hasn't, we really have nothing to do.
-            time.sleep(learner_timeout)
-
             delta = maya.now() - start_time
             if delta.total_seconds() >= timeout:
                 still_checking = ', '.join(to_check)
                 raise RuntimeError(f"Timed out after {timeout} seconds; "
                                    f"need {quantity_remaining} more, still checking {still_checking}.")
 
+            self.alice.block_until_specific_nodes_are_known(to_check, learn_on_this_thread=discover_on_this_thread)
```

Comment:
I don't think the new logic of `sample_essential()` is equivalent to the current state. The original idea was to wait a little to give the nodes a chance to respond, and if some didn't, draw more nodes from the reservoir. The new logic would wait on the original set of nodes until the full timeout expires.

Reply:
Let's make an Issue for this and address it post-merge?

Reply:
Ok, filed issue #2201